central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,530 @@
+	s.schedulerLock.Lock()
+	defer s.schedulerLock.Unlock()
+	s.runningReportConfigIDs.Remove(configID)
+}
+
+func (s *scheduler) UpsertReportSchedule(reportConfig *storage.ReportConfiguration) error {
+	s.cronJobsLock.Lock()
+	defer s.cronJobsLock.Unlock()
+
+	cronSpec := ""
+	var err error
+	var entryID cron.EntryID
+	if reportConfig.GetSchedule() != nil {
+		cronSpec, err = schedule.ConvertToCronTab(reportConfig.GetSchedule())
+		if err != nil {
+			return err
+		}
+		entryID, err = s.cron.AddFunc(cronSpec, s.reportClosure(reportConfig))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Remove the old entry if this is an update
+	if oldEntryID, ok := s.reportConfigToEntryIDs[reportConfig.GetId()]; ok {
+		s.cron.Remove(oldEntryID)
+	}
+	// Only add to entries if this is still a scheduled report config
+	if cronSpec != "" {
+		s.reportConfigToEntryIDs[reportConfig.GetId()] = entryID
+	}
+	return nil
+}
+
+func (s *scheduler) RemoveReportSchedule(reportConfigID string) {
+	s.cronJobsLock.Lock()
+	defer s.cronJobsLock.Unlock()
+
+	oldEntryID, exists := s.reportConfigToEntryIDs[reportConfigID]
+	if exists {
+		s.cron.Remove(oldEntryID)
+		delete(s.reportConfigToEntryIDs, reportConfigID)
+	}
+}
+
+/* Functions to add/remove report jobs from queue */
+
+// CancelReport cancels a report that is still waiting in queue.
+// If the report is already being prepared or has completed execution, it cannot be cancelled.
+func (s *scheduler) CancelReport(ctx context.Context, reportID string) error {
+	metadata, found, err := s.reportMetadataStore.Get(ctx, reportID)
+	if err != nil {
+		return errors.Errorf("Error finding report ID '%s': %s", reportID, err)
+	}
+	if !found {
+		return errors.Errorf("Report ID '%s' does not found", reportID)
+	}
+	if metadata.ReportStatus.RunState == storage.ReportStatus_SUCCESS ||
+		metadata.ReportStatus.RunState == storage.ReportStatus_FAILURE {
+		return errors.Errorf("Cannot cancel. Report ID '%s' has already completed execution", reportID)
+	}
+
+	if !s.tryRemoveReportFromQueue(reportID) {
+		return errors.Errorf("Canot cancel. Report ID '%s' is already being prepared", reportID)
+	}
+	err = s.reportMetadataStore.DeleteReportMetadata(ctx, reportID)
+	if err != nil {
+		return errors.Errorf("Error deleting report ID '%s' from storage", reportID)
+	}
+
+	return nil
+}
+
+func (s *scheduler) tryRemoveReportFromQueue(reportID string) bool {
+	s.schedulerLock.Lock()
+	defer s.schedulerLock.Unlock()
+
+	idxToRemove := -1
+	for i, req := range s.reportsQueue {
+		if req.ReportMetadata.GetReportId() == reportID {
+			idxToRemove = i
+			break
+		}
+	}
+	if idxToRemove < 0 {
+		return false
+	}
+	s.reportsQueue = slices.Delete(s.reportsQueue, idxToRemove, idxToRemove+1)
+	return true
+}
+
+// SubmitReport submits a report for scheduling either on demand or on a schedule.
+// If it's on demand, it will begin running if there isn't already one running from the same requesting user.
```

Comment:
Is this true?  Looks like it just puts it in the queue with all the other reports where it will run if we aren't already running our semaphore max.

Reply:
Yeah, on [ln388](https://github.com/stackrox/stackrox/pull/6818/files#diff-915c4362647c469ec277f2f9a38d9ba728d9eeaefb6f8b814085018808a03efcR388) we check if user has a report already in queue, and error out if there is one.